pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/manager/RedirectManager.java

```
@@ -0,0 +1,114 @@
+            Map<String, BrokerLookupData> map = new ConcurrentHashMap<>();
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+            for (String brokerId : availableBrokers) {
+                futures.add(this.brokerLookupDataLockManager.readLock(
+                        String.format("%s/%s", LOADBALANCE_BROKERS_ROOT, brokerId)).thenAccept(lookupDataOpt -> {
+                    if (lookupDataOpt.isPresent()) {
+                        map.put(brokerId, lookupDataOpt.get());
+                    } else {
+                        log.warn("Got an empty lookup data, brokerId: {}", brokerId);
+                    }
+                }));
+            }
+
+            return FutureUtil.waitForAll(futures).thenApply(__ -> map);
+        });
+    }
+
+    public CompletableFuture<Optional<LookupResult>> findRedirectLookupResultAsync() {
+        String currentLMClassName = pulsar.getConfiguration().getLoadManagerClassName();
+        boolean debug = ExtensibleLoadManagerImpl.debug(pulsar.getConfig(), log);
+        return getAvailableBrokerLookupDataAsync().thenApply(lookupDataMap -> {
+            if (lookupDataMap.isEmpty()) {
+                log.warn("No available broker found");
+                return Optional.empty();
+            }
+            AtomicReference<ServiceLookupData> latestServiceLookupData = new AtomicReference<>();
+            AtomicLong lastStartTimestamp = new AtomicLong(0L);
+            lookupDataMap.forEach((key, value) -> {
+                if (lastStartTimestamp.get() < value.getStartTimestamp()) {
+                    lastStartTimestamp.set(value.getStartTimestamp());
+                    latestServiceLookupData.set(value);
+                }
+            });
+            if (latestServiceLookupData.get() == null) {
+                log.warn("No available broker found");
+                return Optional.empty();
```

Maybe we can throw an exception directly.

---

Updated.